### PR TITLE
study(color): close the two wide-hull gaps the P7 report left open (#4041)

### PR DESCRIPTION
**Pull request by zackees**
`docs/color-gamut-algorithm-selection.md` recorded two things it explicitly could not claim for devices with a white emitter. This measures both.

## 1. Ray connectivity above three emitters

The report establishes that the feasible chroma ray is a single interval — which is what makes the shipped eight-halving bisection a valid search rather than a lucky one — and then says in as many words:

> That is strong evidence for this device, not a proof, and **it says nothing about the >=4-emitter case where the zonotope gains a redundant generator.**

A white emitter changes what feasibility *means*: the preimage of a target stops being unique, so the question goes from "does the one solution land in the box" to "does **some** solution" — which is the predicate `mapAndAllocateRgbwQ16` / `mapAndAllocateRgbwwQ16` actually consult.

Swept at the same density as the three-emitter case, per white-emitter device the corpus ships:

| device | rays | evaluations | disconnected | not anchored at zero |
| --- | --- | --- | --- | --- |
| `rgbw` | 4 680 | 1 881 360 | 0 | 0 |
| `non_d65_white` | 4 680 | 1 881 360 | 0 | 0 |
| `rgbww` | 4 680 | 1 881 360 | 0 | 0 |
| `rgbww_two_white` | 4 680 | 1 881 360 | 0 | 0 |

**7.5 million evaluations, no disconnected interval.** The last column is stated separately because an interval that is connected but floats off the axis would defeat the shipped search just as thoroughly — it seeds its bracket at chroma zero.

The sweep runs against the **enumeration oracle**, not `allocate_two_white`. A sweep built on the closed form could only ever confirm the closed form's own idea of the hull, which is the assumption under test.

## 2. The map-then-allocate composition

The report's "Not covered" section, in full:

> The interaction between the allocation policy and the gamut mapping. Both are characterized here in isolation; a target that is out of gamut *and* on a device with a white emitter goes through both, and that composition has not been scored.

Scored over ~28 000 out-of-gamut targets at the shipped eight halvings, with the drives **re-rendered** rather than trusted — which is what would catch an allocation reporting success while producing something else:

| device | out-of-gamut | hue drift | dE2000, realized vs mapped | worst step | worst / mean |
| --- | --- | --- | --- | --- | --- |
| `rgbw` | 7 012 | 0.000000° | 8.6e-14 | 0.002871 | 12.5x |
| `non_d65_white` | 7 012 | 0.000000° | 7.9e-14 | 0.003140 | 12.7x |
| `rgbww` | 7 012 | 0.000000° | 7.9e-14 | 0.003441 | 10.1x |
| `rgbww_two_white` | 7 441 | 0.000000° | 1.0e-13 | 0.011086 | 9.7x |

Allocation adds nothing measurable, and the worst/mean ratios sit **below** the three-emitter path's 42.8x — the white emitter does not make the boundary step worse.

## Two findings along the way

**The tolerance skew is a trap.** Scoring this the obvious way — oracle in front, closed form behind — reports 56 failures on `rgbww` that *neither method commits alone*. `most_white_two` allows `1e-7`, `allocate_two_white` allows `1e-9`, so hull-boundary targets are accepted by one and refused by the other. Worst case: a blue drive at **-7.6e-08** with the whites off, genuinely outside by less than the oracle's slack. Not a defect — but the mapper and the allocation behind it must consult the *same* predicate, which is what the embedded path does. With both on the closed form the count is 0 everywhere.

**#4245 now has numbers for wide hulls.** Constructing over-bright targets showed the hull reaching well above the brightest *neutral* off the neutral axis:

| device | brightest neutral | brightest reachable | at | headroom |
| --- | --- | --- | --- | --- |
| `rgbw` | 1.338544 | 1.623559 | hue 300, C 0.48 | **+21.3%** |
| `non_d65_white` | 1.333993 | 1.633584 | hue 300, C 0.50 | **+22.5%** |
| `rgbww` | 1.499835 | 1.748872 | hue 300, C 0.46 | **+16.6%** |
| `rgbww_two_white` | 1.152403 | 1.167419 | hue 0, C 0.02 | +1.3% |

Up to **22% of reachable lightness discarded** by the clamp. Not fixed here — the clamp is what makes the chroma bisection valid at all, and removing it without replacing the search inverts the bisection's invariant (measured and written up on #4245).

## Mutation testing

Dropping the lightness clamp from the mapper **initially survived the entire suite.** The grids stop at L = 0.9 and every wide device caps its neutral above 1.15, so no over-bright target was ever generated — the exact gap the report records for the three-emitter corpus ("an over-bright target has to be constructed"). The added case closes it:

| mutation | before | after |
| --- | --- | --- |
| drop the lightness clamp | **survived** | fails on all 4 devices |
| interval detector never closes an interval | fails hole-detection control | — |

The hole-detection control exists precisely so the connectivity sweep cannot pass by never firing: a predicate with a hole punched in it must be reported as two intervals.

## Verification

- `uv run pytest ci/tests/test_color_wide_hull_study.py` — 14 passed, 36 subtests, 3.3s
- `uv run pytest ci/tests/` — 1277 passed, 41 skipped, 2438 subtests. The 2 failures on this branch are the pre-existing ones fixed by #4249, which is not merged yet.
- `bash lint` — clean, exit 0

Refs #4041

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01MLhWkMfzLjrnLTDMBE6Fj9

## Summary by CodeRabbit

* **New Features**
  * Added a study tool for evaluating wide-gamut color devices with four or five emitters.
  * Added analysis of feasible color ranges, gamut connectivity, hue-preserving mapping, allocation accuracy, and reachable brightness.

* **Tests**
  * Added regression coverage for wide-gamut mapping, allocation fidelity, hue preservation, bounds, and edge cases.

* **Documentation**
  * Expanded gamut algorithm documentation with empirical results for wide white-emitter devices and updated known limitations.

### CODE REVIEW

File: ci/color_wide_hull_study.py [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+            low = chroma
+        else:
+            high = chroma
+    return _xyz_from_oklab((lightness, low * math.cos(hue), low * math.sin(hue)))
+
+
+@typechecked
+def realize_one_white(
+    forward: Matrix3, inverse: Matrix3, white: Xyz, target: Xyz
+) -> tuple[Xyz, tuple[float, float, float, float]] | None:
```
[the remark this answers is not in the document]
Comment (zackees) [reply ci/color_wide_hull_study.py]: Valid — done. All three now return a `Realization` with named `realized` and `drives`.

It paid for itself beyond the guideline: `WideDevice.realize` was re-wrapping the pair purely to widen the drives tuple from a fixed arity to `tuple[float, ...]`, and that disappears. The tests also read better — they now `assertIsNotNone` and then access fields, instead of destructuring a value that may be `None`.

14 tests, 36 subtests still pass.

File: docs/color-gamut-algorithm-selection.md [hunk 2]
```diff
@@ -511,9 +516,107 @@ matter:
...
+device the corpus ships, using the *oracle* hull rather than the closed form.
+That choice matters: a sweep built on `allocate_two_white` could only confirm
+that function's own idea of the hull, which is the assumption under test.
+
+| device | rays | evaluations | disconnected | not anchored at zero |
+| --- | --- | --- | --- | --- |
+| `rgbw` | 4 680 | 1 881 360 | 0 | 0 |
+| `non_d65_white` | 4 680 | 1 881 360 | 0 | 0 |
+| `rgbww` | 4 680 | 1 881 360 | 0 | 0 |
+| `rgbww_two_white` | 4 680 | 1 881 360 | 0 | 0 |
```
[the remark this answers is not in the document]
Comment (zackees) [reply docs/color-gamut-algorithm-selection.md:546]: Good catch on the inconsistency, and it resolves the way your second option suggests — the table is right and the prose was not.

`scan_ray` walks `range(samples + 1)`, so 401 counts chroma **steps** and both endpoints are evaluated: 402 points per ray. I verified that rather than reasoning about it, by instrumenting the sweep with a counting predicate:

```
evaluations for samples=401: 402
4680 rays x that = 1881360
four devices     = 7525440
```

So 1 881 360 per device and the 7.5 million total both stand, including the cross-reference at line 226. The text now states the extra endpoint explicitly instead of leaving you to reconcile it:

> The steps are intervals, not points: both endpoints are evaluated, so a ray costs 402 evaluations and not 401.

File: docs/color-gamut-algorithm-selection.md [hunk 3]
```diff
@@ -511,9 +516,107 @@ matter:
...
+| device | brightest neutral | brightest reachable | at | headroom |
+| --- | --- | --- | --- | --- |
+| `rgbw` | 1.338544 | 1.623559 | hue 300, C 0.48 | **+21.3%** |
+| `non_d65_white` | 1.333993 | 1.633584 | hue 300, C 0.50 | **+22.5%** |
+| `rgbww` | 1.499835 | 1.748872 | hue 300, C 0.46 | **+16.6%** |
+| `rgbww_two_white` | 1.152403 | 1.167419 | hue 0, C 0.02 | +1.3% |
+
+So a chromatic target just above the neutral cap can be exactly reachable and
+is compressed anyway. That is #4245, and these are the first numbers on it for
+wide hulls -- up to **22% of reachable lightness** discarded on a four-emitter
```
[the remark this answers is not in the document]
Comment (zackees) [reply docs/color-gamut-algorithm-selection.md]: You're right, and the arithmetic checks out both ways:

```
headroom above cap : 22.46%
share of reachable : 18.34%
```

I quoted the table's number with the other fraction's wording, which is wrong whichever one I meant.

Rather than picking one I've given both, because they answer different questions and neither substitutes for the other — headroom is what the clamp gives up as a fraction of what it keeps, share is the same interval as a fraction of the whole reachable range. The table gains a `share of reachable` column (17.6% / 18.3% / 14.2% / 1.3%) and the sentence now names which is which:

> the clamp gives up **22.5% on top of the neutral cap**, which is **18.3% of the reachable range**

I'd also posted the loose framing on #4245, so I've corrected it there too rather than leaving the wrong denominator standing on the issue.
